spanner/api/Spanner/Program.cs

```diff
@@ -1757,35 +1757,44 @@ await connection.BeginTransactionAsync())
                 $"Data Source=projects/{projectId}/instances/{instanceId}"
                 + $"/databases/{databaseId}";
-            List<Singer> singers = new List<Singer> {
-                new Singer {SingerId = firstSingerId, FirstName = "Marc",
-                    LastName = "Richards"},
-                new Singer {SingerId = secondSingerId, FirstName = "Catalina",
-                    LastName = "Smith"},
-                new Singer {SingerId = 3, FirstName = "Alice",
-                    LastName = "Trentor"},
-                new Singer {SingerId = 4, FirstName = "Lea",
-                    LastName = "Martin"},
-                new Singer {SingerId = 5, FirstName = "David",
-                    LastName = "Lomond"},
+            List<Album> albums = new List<Album> {
+                new Album { SingerId = secondSingerId, AlbumId = 1, AlbumTitle = "Green" },
+                new Album { SingerId = secondSingerId, AlbumId = 3, AlbumTitle = "Terrified" },
             };
             // Create connection to Cloud Spanner.
             using (var connection = new SpannerConnection(connectionString))
             {
                 await connection.OpenAsync();
 
-                // Delete rows from the Singers table.
-                var cmd = connection.CreateDeleteCommand("Singers",
+                // Delete individual rows from the Albums table.
+                var cmd = connection.CreateDeleteCommand(
+                    "Albums",
                     new SpannerParameterCollection {
-                        {"SingerId", SpannerDbType.Int64}
-                    });
-                await Task.WhenAll(singers.Select(singer =>
+                        {"SingerId", SpannerDbType.Int64},
+                        {"AlbumId", SpannerDbType.Int64}
+                    }
+                );
+                await Task.WhenAll(albums.Select(album =>
                 {
-                    cmd.Parameters["SingerId"].Value = singer.SingerId;
+                    cmd.Parameters["SingerId"].Value = album.SingerId;
+                    cmd.Parameters["AlbumId"].Value = album.AlbumId;
                     return cmd.ExecuteNonQueryAsync();
                 }));
 
-                Console.WriteLine("Deleted data.");
+                Console.WriteLine("Deleted individual rows in Albums.");
+
+                // Delete a range of rows from the Singers table where the column key is >=3 and <5.
+                cmd = connection.CreateDmlCommand(
+                   "DELETE FROM Singers WHERE SingerId >= 3 AND SingerId < 5");
+                int rowCount = await cmd.ExecuteNonQueryAsync();
+                Console.WriteLine($"{rowCount} row(s) deleted from Singers.");
+
+                // Delete remaining Singers rows, which will also delete the remaining Albums rows
+                // since it was defined with ON DELETE CASCADE.
+                cmd = connection.CreateDmlCommand(
+                   "DELETE FROM Singers WHERE true");
```

Comment:
Consider removing the "WHERE true", I don't think it's needed.

Reply:
The WHERE clause is unfortunately required for DELETEs in Cloud Spanner. This is to prevent accidentally deleting all data when you meant to delete only a few rows.

https://cloud.google.com/spanner/docs/dml-syntax#where_clause

Reply:
Ah makes sense.